### PR TITLE
ARROW-13399: [R] Update dataset.Rmd vignette

### PR DESCRIPTION
**Pull request by thisisnic**
Various updates to dataset.Rmd including:
* separating out dense text chunks
* rephrasing based on suggestions by Grammarly to simplify phrasing
* rephrasing "we" to "you"

### CODE REVIEW

File: r/vignettes/dataset.Rmd [hunk 1]
```diff
@@ -20,34 +20,36 @@ and what is on the immediate development roadmap.
...
 See `vignette("install", package = "arrow")` for details.
-To see if your `arrow` installation has S3 support, run
+To see if your __arrow__ installation has S3 support, run:
 
 ```{r}
 arrow::arrow_with_s3()
 ```
 
-Even with S3 support enabled network, speed will be a bottleneck unless your
+Even with an S3 support enabled network, speed will be a bottleneck unless your
```
Comment (nealrichardson) [inline r/vignettes/dataset.Rmd:39]:
```suggestion
Even with S3 support enabled, network speed will be a bottleneck unless your
```

**Comment by thisisnic** [comment]
I've also removed a lot of the backticks around "arrow" and "dplyr" as I think they make the text harder to scan.  I took a look at the tidyverse packages and the convention in their vignettes seems to be to link to any external packages the first time they're mentioned, and then subsequently just treat the package names as if they are words.

File: r/vignettes/dataset.Rmd [hunk 2]
```diff
@@ -313,27 +330,29 @@ instead of a file path, or simply concatenate them like `big_dataset <- c(ds1, d
...
 efficient binary columnar format like Parquet or Feather and partitioning based on
-columns commonly used for filtering. However, we don't always get our data delivered
-to us that way. Sometimes we start with one giant CSV. Our first step in analyzing data
+columns commonly used for filtering. However, data isn't always stored that way.
+Sometimes you might start with one giant CSV. The first step in analyzing data 
 is cleaning is up and reshaping it into a more usable form.
 
-The `write_dataset()` function allows you to take a Dataset or other tabular data object---an Arrow `Table` or `RecordBatch`, or an R `data.frame`---and write it to a different file format, partitioned into multiple files.
+The `write_dataset()` function allows you to take a Dataset or another tabular 
+data object - an Arrow Table or RecordBatch, or an R data frame - and write
```
Comment (jonkeane) [inline r/vignettes/dataset.Rmd:338]:
```suggestion
data object - an Arrow Table or RecordBatch, or an R data.frame - and write
```

This is another we should probably standardize on. This isn't perfect, but as a metric: I see 47 results for `data.frame` in arrow/r/man but only 5 results for `data frame`

**Review by jonkeane — COMMENTED** [review]
This looks like a really nice enhancement for readability and style. I have a few comments (and a number of them are "we should standardise this!"), so feel free to take them, discard them, or punt on them

File: r/vignettes/dataset.Rmd [hunk 3]
```diff
@@ -8,10 +8,10 @@ vignette: >
 ---
 
 Apache Arrow lets you work efficiently with large, multi-file datasets.
-The `arrow` R package provides a `dplyr` interface to Arrow Datasets,
```
[the remark this answers is not in the document]
Comment (thisisnic) [reply r/vignettes/dataset.Rmd:11]: I think your suggestions of backticks+curly brackets around package names could be a good shout as it just makes it that tiny bit more scannable than just backticks and would help us distinguish between Arrow (the project and/or C++ implementation) and arrow (the R package).

File: r/vignettes/dataset.Rmd [hunk 4]
```diff
@@ -77,39 +79,44 @@ feel free to grab only a year or two of data.
...
+Other supported formats include: 
+
+* `"feather"` or `"ipc"` (aliases for `"arrow"`, as Feather v2 is the Arrow file format)
+* `"csv"` (comma-delimited files) and `"tsv"` (tab-delimited files)
+* `"text"` (generic text-delimited files - use the `delimiter` argument to specify which to use)
 
-The `partitioning` argument lets us specify how the file paths provide information
-about how the dataset is chunked into different files. Our files in this example
+For text files, you can pass any parsing options (`delim`, `quote`, etc.) to 
+`open_dataset()` that you would otherwise pass to `read_csv_arrow()`.
```
[the remark this answers is not in the document]
Comment (thisisnic) [reply r/vignettes/dataset.Rmd:116]: I have no recollection of that but I'll have a search on JIRA and see what I can find

File: r/vignettes/dataset.Rmd [hunk 5]
```diff
@@ -118,12 +125,12 @@ have file paths like
 ...
 ```
 
-By providing a character vector to `partitioning`, we're saying that the first
-path segment gives the value for `year` and the second segment is `month`.
+By providing a character vector to `partitioning`, you're saying that the first
```
[the remark this answers is not in the document]
Comment (thisisnic) [reply r/vignettes/dataset.Rmd:128]: Great suggestion, being explicit about this can only help understanding.

File: r/vignettes/dataset.Rmd (hunk 3 above)
[the remark this answers is not in the document]
Comment (thisisnic) [reply r/vignettes/dataset.Rmd:11]: Actually; just curly braces or curly braces + backticks?

File: r/vignettes/dataset.Rmd (hunk 4 above)
[the remark this answers is not in the document]
Comment (thisisnic) [reply r/vignettes/dataset.Rmd:116]: Had a look in the docs and I found this in `open_dataset()`: 
> additional arguments passed to dataset_factory() when sources is a directory path/URI or vector of file paths/URIs

Is this what you meant, as in it only works for certain values of `sources`, or something else?

File: r/vignettes/dataset.Rmd (hunk 5 above)
[the remark this answers is not in the document]
Comment (thisisnic) [reply r/vignettes/dataset.Rmd:128]: Done

File: r/vignettes/dataset.Rmd [hunk 6]
```diff
@@ -118,12 +125,12 @@ have file paths like
...
+path segment gives the value for `year`, and the second segment is `month`.
 Every row in `2009/01/data.parquet` has a value of 2009 for `year`
-and 1 for `month`, even though those columns may not actually be present in the file.
+and 1 for `month`, even though those columns may not be present in the file.
 
-Indeed, when we look at the dataset, we see that in addition to the columns present
+Indeed, when you look at the dataset, you can see that in addition to the columns present
 in every file, there are also columns `year` and `month`.
```
[the remark this answers is not in the document]
Comment (thisisnic) [reply r/vignettes/dataset.Rmd:134]: Done

File: r/vignettes/dataset.Rmd [hunk 7]
```diff
@@ -159,37 +166,37 @@ See $metadata for additional Schema metadata
...
-your query on an Arrow Dataset, the `arrow` package raises an error. However,
-for `dplyr` queries on `Table` objects (which are typically smaller in size) the
-package automatically calls `collect()` before processing that `dplyr` verb.
+Suppose you attempt to call unsupported dplyr verbs or unimplemented functions
+in your query on an Arrow Dataset. In that case, the arrow package raises an error. However,
+for dplyr queries on Arrow Table objects (typically smaller in size than Datasets), the
+package automatically calls `collect()` before processing that dplyr verb.
 
-Here's an example. Suppose I was curious about tipping behavior among the
+Here's an example. Suppose that you are curious about tipping behavior among the
```
[the remark this answers is not in the document]
Comment (thisisnic) [reply r/vignettes/dataset.Rmd]: Done

File: r/vignettes/dataset.Rmd (hunk 2 above)
Comment (nealrichardson) [reply r/vignettes/dataset.Rmd:338]: Here's how I think about it when writing: "`data.frame`" is a proper noun, meaning literally that class of R object, while "data frame" is the more conceptual noun. In some cases, one or the other is obviously correct; in others, either might work.

File: r/vignettes/dataset.Rmd (hunk 4 above)
[the remark this answers is not in the document]
Comment (jonkeane) [reply r/vignettes/dataset.Rmd:116]: I was thinking about https://github.com/apache/arrow/blob/master/r/R/dataset-format.R#L135-L153 bit. 

Running the code in the middle of that, I get:

```
>   # Catch any readr-style options specified with full option names that are
>   # supported by read_delim_arrow() (and its wrappers) but are not yet
>   # supported here
>   unsup_readr_opts <- setdiff(
+     names(formals(read_delim_arrow)),
+     names(formals(readr_to_csv_parse_options))
+   )
> unsup_readr_opts
 [1] "file"              "schema"            "col_names"         "col_types"         "col_select"       
 [6] "na"                "quoted_na"         "skip"              "parse_options"     "convert_options"  
[11] "read_options"      "as_data_frame"     "timestamp_parsers"
```

[the remark this answers is not in the document]
Comment (nealrichardson) [reply r/vignettes/dataset.Rmd:116]: you may want to intersect that with `names(formals(readr::read_delim))` since some of those are arrow function args

File: r/vignettes/dataset.Rmd (hunk 2 above)
Comment (thisisnic) [reply r/vignettes/dataset.Rmd:338]: How about in this example here @nealrichardson ? I was thinking `data frame` as we're referring to both `data.frame` and `tibble::tibble` objects but then they're both of or inherit from class `data.frame ` so wasn't sure what makes most sense.

Comment (thisisnic) [reply r/vignettes/dataset.Rmd:338]: This is an example of both I found and was trying to emulate: https://r4ds.had.co.nz/tibbles.html

Comment (nealrichardson) [reply r/vignettes/dataset.Rmd:338]: I know 😞  I think I'd go with "data frame" so as to (in theory) avoid the "you said data.frame but it's a tibble" though my reasoning for picking one or the other is more about what's in my head when I'm writing, and you're right, the distinction is not perfect in the real world, nor may the reader share my distinction.

File: r/vignettes/dataset.Rmd (hunk 3 above)
[the remark this answers is not in the document]
Comment (thisisnic) [reply r/vignettes/dataset.Rmd:11]: Braces more of a Twitter/blog post thing - PR now updated with `STYLE.md` which contains some guidelines on this and suggests using the tidyverse style of things.

File: r/vignettes/dataset.Rmd (hunk 4 above)
[the remark this answers is not in the document]
Comment (thisisnic) [reply r/vignettes/dataset.Rmd:116]: I'm confused by @nealrichardson 's comment, please can you rephrase that?

In the meantime, it looks like there are only 5 parsing options supported and more unsupported, so I feel like I'm better off explicitly listing the ones that are supported.

File: r/vignettes/dataset.Rmd (hunk 2 above)
Comment (thisisnic) [reply r/vignettes/dataset.Rmd:345]: I'm leaning towards this interpretation too, so have added guidance in `STYLE.md`.

File: r/vignettes/dataset.Rmd [hunk 8]
```diff
@@ -8,10 +8,10 @@ vignette: >
...
 Apache Arrow lets you work efficiently with large, multi-file datasets.
-The `arrow` R package provides a `dplyr` interface to Arrow Datasets,
-as well as other tools for interactive exploration of Arrow data.
+The arrow R package provides a [dplyr](https://dplyr.tidyverse.org/) interface to Arrow Datasets,
+and other tools for interactive exploration of Arrow data.
 
-This vignette introduces Datasets and shows how to use `dplyr` to analyze them.
+This vignette introduces Datasets and shows how to use dplyr to analyze them.
 It describes both what is possible to do with Arrow now
 and what is on the immediate development roadmap.
```
Comment (thisisnic) [inline r/vignettes/dataset.Rmd:16]: Does it? I think we should delete this sentence but it'd be good to hear other people's thoughts on this first.

File: r/vignettes/dataset.Rmd (hunk 4 above)
[the remark this answers is not in the document]
Comment (nealrichardson) [reply r/vignettes/dataset.Rmd]: Many of the args in `unsup_readr_opts` aren't actually from readr, they're just arguments to read_delim_arrow that aren't in readr_to_csv_parse_options. We only care about the ones that are readr options, so:

```
> intersect(unsup_readr_opts, names(formals(readr::read_delim)))
[1] "file"       "col_names"  "col_types"  "col_select" "na"        
[6] "quoted_na"  "skip" 
```

File: r/vignettes/dataset.Rmd (hunk 8 above)
Comment (nealrichardson) [reply r/vignettes/dataset.Rmd:16]: At one point this was true (the discussion at the end talked about what's not yet implemented but coming) but perhaps that's no longer accurate.

File: r/STYLE.md [hunk 9]
```diff
@@ -0,0 +1,19 @@
+# Style
```
Comment (nealrichardson) [inline r/STYLE.md:1]: Need to add the ASF license header here

File: r/vignettes/dataset.Rmd (hunk 4 above)
[the remark this answers is not in the document]
Comment (thisisnic) [reply r/vignettes/dataset.Rmd]: Right, I get you, I think. I think that now I'm explicitly specifying the arguments that *can* be passed through rather than *can't*, I don't think I need to make any more changes here on account of the above comments?

File: r/STYLE.md [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+
+This is a style guide to writing documentation for arrow.
+
+## Coding style
+
+Please use the [tidyverse coding style](https://style.tidyverse.org/).
+
+## Referring to external packages
+
+When referring to external packages, include a link to the package at the first mention, and subsequently refer to it in plain text, e.g.
```
Comment (nealrichardson) [inline r/STYLE.md:30]:
```suggestion
When referring to external packages in documentation, include a link to the package at the first mention, and subsequently refer to it in plain text, e.g.
```

File: r/STYLE.md [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+
+* "The arrow R package provides a [dplyr](https://dplyr.tidyverse.org/) interface to Arrow Datasets.  This vignette introduces Datasets and shows how to use dplyr to analyze them."
+
+## Data frames
+
+When referring to the concept, use the phrase "data frame", whereas when referring to an object of that class or when the class is important, write `data.frame`, e.g.
+
+* "You can call `write_dataset()` on tabular data objects such as Arrow Tables or RecordBatchs, or R data frames. If working with data frames you might want to use a `tibble` instead of a `data.frame` to take advantage of the default behaviour of partitioning data based on grouped variables."
```
Comment (nealrichardson) [inline r/STYLE.md:38]:
```suggestion
* "You can call `write_dataset()` on tabular data objects such as Arrow Tables or RecordBatches, or R data frames. If working with data frames you might want to use a `tibble` instead of a `data.frame` to take advantage of the default behaviour of partitioning data based on grouped variables."
```

File: r/vignettes/dataset.Rmd [hunk 12]
```diff
@@ -159,37 +171,37 @@ See $metadata for additional Schema metadata
...
+you could have just called `ds <- open_dataset("nyc-taxi")` and the partitions
 would have been detected automatically.
 
 ## Querying the dataset
 
-Up to this point, we haven't loaded any data: we have walked directories to find
-files, we've parsed file paths to identify partitions, and we've read the
-headers of the Parquet files to inspect their schemas so that we can make sure
-they all line up.
+Up to this point, you haven't loaded any data.  You've walked directories to find
```
Comment (nealrichardson) [inline r/vignettes/dataset.Rmd:188]:
```suggestion
Up to this point, you haven't loaded any data. You've walked directories to find
```

File: r/vignettes/dataset.Rmd [hunk 13]
```diff
@@ -159,37 +171,37 @@ See $metadata for additional Schema metadata
...
 or other verbs with aggregate functions, use `collect()` to pull the selected
 subset of the data into an in-memory R data frame.
 
-If you attempt to call unsupported `dplyr` verbs or unimplemented functions in
-your query on an Arrow Dataset, the `arrow` package raises an error. However,
-for `dplyr` queries on `Table` objects (which are typically smaller in size) the
-package automatically calls `collect()` before processing that `dplyr` verb.
+Suppose you attempt to call unsupported dplyr verbs or unimplemented functions
+in your query on an Arrow Dataset. In that case, the arrow package raises an error. However,
+for dplyr queries on Arrow Table objects (typically smaller in size than Datasets), the
```
Comment (nealrichardson) [inline r/vignettes/dataset.Rmd:201]:
```suggestion
for dplyr queries on Arrow Table objects (which are already in memory), the
```

File: r/vignettes/dataset.Rmd [hunk 14]
```diff
@@ -228,12 +240,11 @@ cat("
 ")
 ```
 
-We just selected a subset out of a dataset with around 2 billion rows, computed
-a new column, and aggregated on it in under 2 seconds on my laptop. How does
+You've just selected a subset out of a dataset with around 2 billion rows, computed
+a new column, and aggregated it in under 2 seconds on most modern laptops. How does
```
Comment (nealrichardson) [inline r/vignettes/dataset.Rmd:244]:
```suggestion
a new column, and aggregated it in under 2 seconds on a modern laptop. How does
```

File: r/vignettes/dataset.Rmd [hunk 15]
```diff
@@ -259,47 +270,58 @@ See $.data for the source Arrow object
...
 you can build up a query that selects down to a small subset without generating
 intermediate datasets that would potentially be large.
 
 Second, all work is pushed down to the individual data files,
 and depending on the file format, chunks of data within the files. As a result,
-we can select a subset of data from a much larger dataset by collecting the
-smaller slices from each file--we don't have to load the whole dataset in memory
-in order to slice from it.
+you can select a subset of data from a much larger dataset by collecting the
+smaller slices from each file - you don't have to load the whole dataset in
```
Comment (nealrichardson) [inline r/vignettes/dataset.Rmd:281]: em-dash
```suggestion
smaller slices from each file—you don't have to load the whole dataset in 
```

File: r/vignettes/dataset.Rmd [hunk 16]
```diff
@@ -259,47 +270,58 @@ See $.data for the source Arrow object
...
-The schema specification just lets you declare what you want the result to be.
+
+### Work with files in a directory
+
+If you are working with a single file or a set of files that are not all in the 
+same directory, you can provide a file path or a vector of multiple file paths 
+to `open_dataset()`. This is useful if, for example, you have a single CSV file 
+that is too big to read into memory. You could pass the file path to 
+`open_dataset()`, use `group_by()` to partition the Dataset into manageable chunks, 
+then use `write_dataset()` to write each chunk to a separate Parquet file - all
```
Comment (nealrichardson) [inline r/vignettes/dataset.Rmd:302]:
```suggestion
then use `write_dataset()` to write each chunk to a separate Parquet file—all 
```

File: r/vignettes/dataset.Rmd (hunk 2 above)
Comment (nealrichardson) [reply r/vignettes/dataset.Rmd:343]: more em-dashes

```suggestion
data object—an Arrow Table or RecordBatch, or an R data frame—and write
```

**Review by nealrichardson — APPROVED** [review]
Some final nits; LGTM thanks!
